File: iris/monitors/process.py
"""
Process monitor — polls psutil for new or suspicious processes.

Suspicious criteria:
  - Process name in the blacklist (nc, ncat, netcat, socat, mimikatz, …)
  - Executable or CWD under /tmp or /dev/shm
"""

import logging
import queue
import threading
import time

import psutil

from iris.config import Config

logger = logging.getLogger(__name__)

_BLACKLISTED_NAMES: set[str] = {
    "nc", "ncat", "netcat", "socat",
    "mimikatz", "msfconsole", "msfvenom",
}

# Command substrings that indicate downloading to suspicious paths
_SUSPICIOUS_PATH_PREFIXES = ("/tmp/", "/dev/shm/", "/var/tmp/")


def _is_blacklisted(name: str, cmdline: list[str]) -> bool:
    if name.lower() in _BLACKLISTED_NAMES:
        return True
    if name.lower() in ("wget", "curl"):
        return any(arg.startswith(p) for arg in cmdline for p in _SUSPICIOUS_PATH_PREFIXES)
    return False


def _is_suspicious_path(exe: str | None, cwd: str | None) -> bool:
    return any(c and c.startswith(p) for c in (exe, cwd) for p in _SUSPICIOUS_PATH_PREFIXES)


def _snapshot_pids() -> set[int]:
    try:
        return set(psutil.pids())
    except Exception:
        return set()


_ERR = (psutil.NoSuchProcess, psutil.AccessDenied, FileNotFoundError)


def _process_info(proc: psutil.Process) -> dict:
    def _get(fn, default=None):
        try: return fn()
        except _ERR: return default

    return {
        "pid": proc.pid,
        "name": _get(proc.name, "<unknown>"),
        "cmdline": _get(proc.cmdline, []),
        "username": _get(proc.username, "<unknown>"),
        "cwd": _get(proc.cwd),
        "exe": _get(proc.exe),
        "create_time": _get(proc.create_time),
    }


class ProcessMonitor:
    """Polls psutil every interval_seconds, emits events for new suspicious processes."""

    def __init__(self, config: Config, event_queue: queue.Queue) -> None:
        self._config = config
        self._q = event_queue
        self._known_pids: set[int] = _snapshot_pids()
        self._thread: threading.Thread | None = None
        self._stop_event = threading.Event()

    def start(self) -> None:
        self._thread = threading.Thread(target=self._loop, name="iris-process-monitor", daemon=True)
        self._thread.start()
        logger.info("Process monitor started (interval=%ds)", self._config.interval_seconds)

    def stop(self) -> None:
        self._stop_event.set()
        if self._thread:
            self._thread.join(timeout=5)
        logger.info("Process monitor stopped")

    def _loop(self) -> None:
        while not self._stop_event.wait(self._config.interval_seconds):
            self._poll()

    def _poll(self) -> None:
        current_pids = _snapshot_pids()
        new_pids = current_pids - self._known_pids
        self._known_pids = current_pids

        for pid in new_pids:
            try:
                proc = psutil.Process(pid)
                info = _process_info(proc)
            except psutil.NoSuchProcess:
                # Process already exited — probably a very short-lived tool
                continue
            except Exception as exc:
                logger.debug("Failed to inspect PID %d: %s", pid, exc)
                continue

            name = info.get("name", "")
            cmdline = info.get("cmdline") or []
            exe = info.get("exe")
            cwd = info.get("cwd")

            if _is_blacklisted(name, cmdline):
                self._emit("high", f"Blacklisted process detected: {name} (PID {pid})", info)
            elif _is_suspicious_path(exe, cwd):
                self._emit(
                    "medium",
                    f"Process running from suspicious path: {name} (PID {pid})",
                    info,
                )

    def _emit(self, severity: str, title: str, payload: dict) -> None:
        evt = {
            "event_type": "suspicious_process",
            "severity": severity,
            "title": title,
            "payload": payload,
        }
        try:
            self._q.put_nowait(evt)
        except queue.Full:
            logger.warning("Process monitor event queue full, dropping event")

File: iris-rs/src/monitors/process.rs
use serde_json::{json, Value};
use std::collections::HashSet;
use std::time::Duration;
use tokio::sync::mpsc;
use tokio::task::JoinHandle;

use crate::config::Config;

const BLACKLIST: &[&str] = &["nc", "ncat", "netcat", "socat", "mimikatz", "msfconsole", "msfvenom"];
const SUSPICIOUS_PREFIXES: &[&str] = &["/tmp/", "/dev/shm/", "/var/tmp/"];

pub fn spawn(config: Config, tx: mpsc::Sender<Value>) -> JoinHandle<()> {
    tokio::spawn(async move {
        let mut known: HashSet<i32> = snapshot_pids();
        let interval = Duration::from_secs(config.interval_seconds);
        tracing::info!("Process monitor started (interval={}s)", config.interval_seconds);
        loop {
            tokio::time::sleep(interval).await;
            let current = snapshot_pids();
            let new_pids: Vec<i32> = current.difference(&known).copied().collect();
            known = current;
            for pid in new_pids {
                if let Some(evt) = inspect(pid) {
                    if tx.send(evt).await.is_err() { return; }
                }
            }
        }
    })
}

fn snapshot_pids() -> HashSet<i32> {
    procfs::process::all_processes()
        .map(|iter| iter.flatten().map(|p| p.pid).collect())
        .unwrap_or_default()
}

fn inspect(pid: i32) -> Option<Value> {
    let proc = procfs::process::Process::new(pid).ok()?;
    let name = proc.stat().ok().map(|s| s.comm).unwrap_or_default();
    let cmdline = proc.cmdline().unwrap_or_default();
    let exe  = proc.exe().ok().map(|p| p.to_string_lossy().into_owned());
    let cwd  = proc.cwd().ok().map(|p| p.to_string_lossy().into_owned());

    let (severity, title) = if is_blacklisted(&name, &cmdline) {
        ("high", format!("Blacklisted process: {} (PID {})", name, pid))
    } else if suspicious_path(exe.as_deref(), cwd.as_deref()) {
        ("medium", format!("Process from suspicious path: {} (PID {})", name, pid))
    } else {
        return None;
    };

    Some(json!({
        "event_type": "new_process",
        "severity":   severity,
        "title":      title,
        "payload": {
            "pid":      pid,
            "name":     name,
            "cmdline":  cmdline,
            "exe":      exe,
            "cwd":      cwd,
        }
    }))
}

fn is_blacklisted(name: &str, cmdline: &[String]) -> bool {
    let n = name.to_lowercase();
    if BLACKLIST.contains(&n.as_str()) { return true; }
    if n == "wget" || n == "curl" {
        return cmdline.iter().any(|a| SUSPICIOUS_PREFIXES.iter().any(|p| a.starts_with(p)));
    }
    false
}

fn suspicious_path(exe: Option<&str>, cwd: Option<&str>) -> bool {
    [exe, cwd].iter().flatten().any(|p| SUSPICIOUS_PREFIXES.iter().any(|pre| p.starts_with(pre)))
}
